src/shillelagh/adapters/api/google_analytics.py

```diff
@@ -0,0 +1,218 @@
+        self.operation = operation
+        self.parameters = parameters
+        self.date_range = date_range
+
+        credentials = ServiceAccountCredentials.from_json_keyfile_dict(
+            service_account_info, SCOPES
+        )
+        self.analytics_service = build(
+            "analyticsreporting", "v4", credentials=credentials
+        )
+
+        self.parsed_operation = parse(operation)
+        parsed_columns = self.get_query_columns()
+        self._set_columns(parsed_columns)
+
+    def supports(uri: str, fast: bool, **kwargs):
+        if uri.startswith("google_analytics"):
+            return True
+
+    @staticmethod
+    def parse_uri(uri: str) -> Tuple[Any, ...]:
+        return (uri,)
+
+    @staticmethod
+    def need_operation():
+        return True
+
+    def get_query_columns(self):
+        stmt = self.parsed_operation[0]
+        columns = []
+        column_identifiers = []
+
+        # get column_identifieres
+        in_select = False
+        for token in stmt.tokens:
+            if isinstance(token, Comment):
+                continue
+            if str(token).lower() == "select":
+                in_select = True
+            elif in_select and token.ttype is None:
+                if isinstance(token, Identifier):
+                    column_identifiers.append(token)
+                elif isinstance(token, IdentifierList):
+                    for identifier in token.get_identifiers():
+                        column_identifiers.append(identifier)
+                break
+
+        # get column names
+        for column_identifier in column_identifiers:
+            columns.append(column_identifier.get_name().replace('`', ''))
+
+        return columns
+
+    # def get_date_range(self):
+    #     parsed_operation = self.parsed_operation[0]
+    #     for token in parsed_operation:
+    #         if isinstance(token, Where) and token.value.startswith("where dateRange in"):
+    #             range_string = token.value.replace("where dateRange in", "").strip().replace("(", "").replace(")", "").split(",")
+    #             if (len(range_string) == 2):
+    #                 return range_string[0].strip(), range_string[1].strip()
+    #     return None
+
+    def get_data(
+        self, bound: Dict[str, Filter], order: List[Tuple[str, RequestedOrder]]
+    ) -> Iterator[Row]:
+        metrics = []
+        dimensions = []
+
+        for col in self.columns:
+            if col in ALL_DIMENSIONS:
+                dimensions.append({"name": col})
+            elif col in ALL_METRICS:
+                metrics.append({"expression": col})
+
+        # date_range = self.get_date_range()
+        
+        data = (
+            self.analytics_service.reports()
+            .batchGet(
+                body={
+                    "reportRequests": [
+                        {
+                            "viewId": self.view_id,
+                            "dateRanges": self.date_range,
```

Comment:
It would be better to build this from `bounds`. Your adapter should have a date column in its `columns`:

```python
from shillelagh.fields import Date, Order
from shillelagh.filters import Range

columns = {
    "date": Date(filters=[Range], order=Order.NONE, exact=True),
    ...,
}
```

Then in `get_data` you can build `dateRanges` from `bounds`:

```python
date_bounds = bounds.get("date", Range())
default_start = date.today() - timedelta(days=7)
default_end = date.today()
date_range = [
    {
        "startDate": str(date_bounds.start or default_start),
        "endDate" : str(date_bounds.end or default_end),
    },
]
```

Reply:
yess, thanks for this, works perfectly fine.

Reply:
Yeah, the WeatherAPI does something similar: https://github.com/betodealmeida/shillelagh/blob/b3487120a89ab653feb84039f157b1321fe7c89b/docs/adapters.rst#weatherapi

By default it looks at a "window" of 7 days:

https://github.com/betodealmeida/shillelagh/blob/8d2ebd7e267684b20a4cf769cb39d173460ce86e/src/shillelagh/adapters/api/weatherapi.py#L186-L189